Source/WebKit/NetworkProcess/webtransport/cocoa/NetworkTransportSessionCocoa.mm

```diff
@@ -37,121 +37,205 @@
+        if (error)
+            return; // FIXME: Use weakNetworkTransportSession to pipe the failure to JS.
+        switch (state) {
+        case nw_listener_state_waiting:
+            return; // We will get another callback with another state change.
+        case nw_listener_state_ready:
+            return;
+        case nw_listener_state_failed:
+        case nw_listener_state_cancelled:
+        case nw_listener_state_invalid:
+            // FIXME: Use weakNetworkTransportSession to pipe the failure to JS.
+            return;
+        }
+        RELEASE_ASSERT_NOT_REACHED();
+    }).get());
+
+    nw_listener_set_queue(listener.get(), dispatch_get_main_queue());
+    nw_listener_start(listener.get());
+
 }
 
 void NetworkTransportSession::sendDatagram(std::span<const uint8_t> data, CompletionHandler<void()>&& completionHandler)
 {
-    nw_connection_group_send_message(m_connectionGroup.get(), makeDispatchData(Vector(data)).get(), m_endpoint.get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+    nw_connection_send(m_connections.first().get(), makeDispatchData(Vector(data)).get(), NW_CONNECTION_DEFAULT_MESSAGE_CONTEXT, true, makeBlockPtr([completionHandler = WTFMove(completionHandler)] (nw_error_t error) mutable {
+        ASSERT(!error);
         // FIXME: Pass any error through to JS.
         completionHandler();
     }).get());
 }
 
 void NetworkTransportSession::createBidirectionalStream(CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
 {
-    RetainPtr connection = adoptNS(nw_connection_group_extract_connection(m_connectionGroup.get(), m_endpoint.get(), nullptr));
+    createStream(NetworkTransportStreamType::Bidirectional, WTFMove(completionHandler));
+}
+
+void NetworkTransportSession::createOutgoingUnidirectionalStream(CompletionHandler<void(std::optional<WebTransportStreamIdentifier>)>&& completionHandler)
+{
+    createStream(NetworkTransportStreamType::OutgoingUnidirectional, WTFMove(completionHandler));
+}
+
+void NetworkTransportSession::setupListenerConnectionHandler()
+{
+    nw_listener_set_new_connection_handler(m_listener.get(), makeBlockPtr([weakThis = WeakPtr { *this }] (nw_connection_t inboundConnection) mutable {
+        ASSERT_UNUSED(weakThis, weakThis);
+        // FIXME: Implement incoming connection handler
+    }).get());
+}
+
+RetainPtr<nw_parameters_t> NetworkTransportSession::createParameters(NetworkTransportStreamType streamType)
+{
+    ASSERT(streamType != NetworkTransportStreamType::IncomingUnidirectional);
+    RetainPtr<nw_parameters_t> parameters = adoptNS(nw_parameters_copy(nw_connection_copy_parameters(m_connections.first().get())));
+    if (!parameters) {
+        ASSERT_NOT_REACHED();
+        return nil;
+    }
+    nw_protocol_stack_t default_stack = nw_parameters_copy_default_protocol_stack(parameters.get());
```

Comment:
I think this needs adoptNS because nw_parameters_copy_default_protocol_stack  has NW_RETURNS_RETAINED